### PR TITLE
[Draft][luci] infer dynamic shape for pad

### PR DESCRIPTION
**Pull request by icodo98**
This infers dynmic shaep for pad.
If input shape is unknown, output shape is also unknown.

for  #13573

ONE-DCO-1.0-Signed-off-by: JuYoung Lee <rsb98759@gmail.com>

### CODE REVIEW

**Comment by shs-park** [comment]
@Samsung/one_compiler,

This is the first DRAFT from @Samsung/ssafy_2024.
I can't believe just yesterday was the SSAFY Project Starting Ceremony.

We kindly ask for your interest and feedback.
😅

**Comment by shs-park** [comment]
> This infers dynmic shaep for pad.

`dynmic shaep` -> `dynamic shape`
Typo? 😅

If you want to modify the last commit message of yours, 
try using `git commit --amend` and `git push -f [branch] [remote]` forced push option.

**Comment by ragmani** [comment]
> `git push -f [branch] [remote]`

Maybe `git push -f [remote] [branch]`

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 1]
```diff
@@ -31,3 +33,42 @@ TEST(CloneNodeTest, clone_Pad)
   auto cloned_pad = dynamic_cast<luci::CirclePad *>(cloned);
   ASSERT_NE(nullptr, cloned_pad);
 }
+
+TEST(ShapeRuleTest, pad_dynamic_shape)
```
Comment (seanshpark) [inline compiler/luci/service/src/Nodes/CirclePad.test.cpp:37]: plz add negative test with `_NEG` suffix having count(positive) <= count(negative)

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp [hunk 2]
```diff
@@ -237,25 +237,33 @@ loco::NodeShape use_paddings(const CIRCLENODE *node, const luci::CircleConst *pa
...
-      auto pr = paddings->at<S64>(idx + 1);
-      auto max = static_cast<int64_t>(std::numeric_limits<int32_t>::max());
-      auto low = static_cast<int64_t>(std::numeric_limits<int32_t>::lowest());
-      LUCI_ASSERT(pl <= max, "paddings is over 32 bit limit");
-      LUCI_ASSERT(pl >= low, "paddings is over 32 bit limit");
-      LUCI_ASSERT(pr <= max, "paddings is over 32 bit limit");
-      LUCI_ASSERT(pr >= low, "paddings is over 32 bit limit");
-      value += static_cast<int32_t>(pl); // left
-      value += static_cast<int32_t>(pr); // right
+
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInferenceRule.cpp:246]:
```suggestion
```

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp [hunk 3]
```diff
@@ -237,25 +237,33 @@ loco::NodeShape use_paddings(const CIRCLENODE *node, const luci::CircleConst *pa
   {
     int32_t idx = ni * 2;
     int value = input_shape.dim(ni).value();
-    if (paddings->dtype() == S32)
+    if (!input_shape.dim(ni).known())
```
Comment (seanshpark) [inline compiler/luci/service/src/CircleShapeInferenceRule.cpp:240]: (optional)

```cxx

  for (int32_t ni = 0; ni < n; ++ni)
  {
    if (not input_shape.dim(ni).known())
    {
      output_shape.dim(ni).unset();
      continue;
    }
    int32_t idx = ni * 2;
```

**Comment by icodo98** [comment]
> If you want to modify the last commit message of yours, ~try using `git commit --amend` and `git push -f [branch] [remote]` forced push option.~ try using `git commit --amend` and `git push -f [remote] [branch]` forced push option.

Sorry for typo. 😂 I'll fix it.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 1 above)
Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:37]: Just to be clear, do you mean testing for pad inference if it fails?
I don't understand 
> suffix having count(positive) <= count(negative)
this line.

File: compiler/luci/service/src/CircleShapeInferenceRule.cpp (hunk 3 above)
Comment (icodo98) [reply compiler/luci/service/src/CircleShapeInferenceRule.cpp:240]: I thought continue statement might violate convention.😂
If it's fine, this is cleaner code.

**Comment by shs-park** [comment]
@Samsung/ssafy_2024,

Please note this PR #13738.
The code of shape inference is being consolidated into another place.
😅

**Comment by icodo98** [comment]
> Please note this PR #13738. The code of shape inference is being consolidated into another place. 😅
I checked #13738 .  I'll change my code to same place with that PR.

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp (hunk 1 above)
Comment (zetwhite) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:37]: > Just to be clear, do you mean testing for pad inference if it fails?
 
Yes, you're right. 

---------------------------------------

( I put my understanding here ) 

Negative tests mean kinds of these 
  * When unexpected data are given, it throws an error as expected (make sure that the system does not just die)
  * ... :thinking: 

As I know, there is a policy that we should have more negative tests than positive tests. 

You can refer this PR (https://github.com/Samsung/ONE/pull/13698). There are two negative tests and one positive test. Also, negative tests have suffix `_NEG` to mark that it is negative test.

Comment (icodo98) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:37]: > You can refer this PR (#13698). There are two negative tests and one positive test. Also, negative tests have suffix `_NEG` to mark that it is negative test.

Thanks for kind reply. I'll add some negative tests. 
In my opinion, there are some negative cases with pad operations.

&emsp; 1.  paddings shape is not [n,2]
&emsp; 2.   paddings is over 32 bit limit

Is there any other suggestions?

File: compiler/luci/service/src/Nodes/CirclePad.test.cpp [hunk 4]
```diff
@@ -31,3 +33,59 @@ TEST(CloneNodeTest, clone_Pad)
...
+  ASSERT_TRUE(shape.dim(3).known());
+
+  ASSERT_EQ(1, shape.dim(0).value());
+  ASSERT_EQ(2, shape.dim(1).value());
+  ASSERT_EQ(0, shape.dim(2).value());
+  ASSERT_EQ(4, shape.dim(3).value());
+  pad.drop();
+}
+
+TEST(ShapeRuleTest, pad_without_padding_NEG)
```
[the remark this answers is not in the document]
Comment (shs-park) [reply compiler/luci/service/src/Nodes/CirclePad.test.cpp:75]: Now, you can try dividing it into smaller PRs and uploading the actual PR.
In fact, code reviews between developers are done much more on the actual PR.

Good luck 🚀
